docs/source/notes/numpy_compatibility.rst

```diff
@@ -0,0 +1,388 @@
+.. _numpy_compatibility:
+
+.. testsetup:: *
+
+   import numpy as np
+   import torch
+
+NumPy Compatibility
+===================
+
+.. note::
+
+   **TL;DR:**
+
+   - Be explicit about obtaining and using NumPy arrays
+   - `This tutorial`_ shows a more concrete example of interoperability between
+     NumPy and PyTorch
+
+At the onset it should be stressed that PyTorch tensors and NumPy arrays are
+fundamentally different types. However, in the interests of being accessible to
+the wider scientific python ecosystem, there are implicit and explicit
+conversion methods. The major considerations to keep in mind are:
+
+- **Where** is the structure allocated?
+    * This may be on a compute device (e.g. a GPU) or the CPU
+- **What** is the underlying data?
+    * In particular, several functions will either return a copy or a view
+- **How** are modifications to the data handled?
+    * The data returned may be read-only, or may be writable
+
+From the end-user perspective, it is most desirable to have zero copy methods,
+however, for exploratory analyses and their visualizations in particular, it is
+often more important to have guaranteed conversions, even if copies are created.
+
+Basic interoperability
+----------------------
+
+NumPy arrays have no conception of the GPU, and so all conversions require a
+``cpu`` tensor. Similarly, the **computational graph** cannot be represented in
+```numpy``, and so the pre-requisite to have an operation or object without one,
+i.e. only leaf nodes are supported.
+
+.. doctest::
+
+   >>> b_torch = torch.ones(3)
+   >>> assert(b_torch.requires_grad == False)
+   >>> assert(b_torch.is_cuda == False)
+   >>> b_torch.numpy()
+   array([1., 1., 1.], dtype=float32)
+
+PyTorch will raise exceptions instead of allowing operations that may lead to
```

Comment:
This sentence is very general and I'm not sure it's strictly true?

Reply:
Maybe "instead of mixing PyTorch tensors with NumPy arrays in function calls or Python's built-in operators"? I can't immediately think of any exceptions there, and it's by design that only `Tensor`, `Tensor` subclasses, and other explicitly-allowed objects like the ones with `__torch_function__` / `__torch_dispatch__` methods are accepted.